chirp/drivers/kenwood_d7.py

```diff
@@ -13,332 +13,3404 @@
+        if len(spec) == 14:  # This is a Call memory
+            spec.insert(1, 0)   # Add "bank"
+            spec.insert(15, 0)  # Add "skip"
+            memid = self._call_index_to_memid(int(spec[0]))
+            mem.extd_number = memid
+            mem.number = self._memid_to_index(mem.extd_number)
+        elif spec[2].isdigit():  # Normal memory
+            mem.number = int(spec[2])
+            memid = self._index_to_memid(mem.number)
+        else:  # Program Scan memory
+            memid = spec[2]
+            mem.extd_number = memid
+            mem.number = self._memid_to_index(mem.extd_number)
+        mem.immutable = self._get_immutable(memid)
         mem.freq = int(spec[3], 10)
-        mem.tuning_step = STEPS[int(spec[4], 16)]
-        mem.duplex = DUPLEX[int(spec[5])]
-        mem.tmode = get_tmode(spec[7], spec[8], spec[9])
-        mem.rtone = self._kenwood_valid_tones[int(spec[10]) - 1]
-        mem.ctone = self._kenwood_valid_tones[int(spec[12]) - 1]
-        if spec[11] and spec[11].isdigit():
-            mem.dtcs = chirp_common.DTCS_CODES[int(spec[11][:-1]) - 1]
-        elif spec[11] == '':
-            pass
-        else:
-            LOG.warn("Unknown or invalid DCS: %s" % spec[11])
+        mem.tuning_step = self._STEPS[int(spec[4], 16)]
+        mem.duplex = self._DUPLEX[int(spec[5])]
+        mem.tmode = self._get_tmode(spec[7], spec[8], spec[9])
+        mem.rtone = self._TONES[int(spec[10]) - 1]
+        mem.ctone = self._TONES[int(spec[12]) - 1]
         if spec[13]:
             mem.offset = int(spec[13])
         else:
             mem.offset = 0
-        mem.mode = MODES[int(spec[14])]
-        mem.skip = int(spec[15]) and "S" or ""
-
+        mem.mode = self._MODES[int(spec[14])]
+        if 'skip' not in mem.immutable:
+            mem.skip = int(spec[15]) and "S" or ""
+        self._parse_mem_fixup(mem, spec)
         return mem
 
-    EXTRA_BOOL_SETTINGS = {
-        'main': [("LMP", "Lamp")],
-        'dtmf': [("TXH", "TX Hold")],
-    }
-    EXTRA_LIST_SETTINGS = {
-        'main': [("BAL", "Balance"),
-                 ("MNF", "Memory Display Mode")],
-        'save': [("SV", "Battery Save")],
-    }
-
-    def _get_setting_options(self, setting):
-        opts = self._SETTINGS_OPTIONS[setting]
-        if opts is None:
-            opts = getattr(self, '_%s_OPTIONS' % setting)
-        return opts
+    def _parse_split(self, mem, result):
+        value = result.split(" ")[1]
+        spec = value.split(",")
+        mem.duplex = "split"
+        # TODO: There's another parameter here... not sure what it is
+        if len(spec) == 14:  # Call memory
+            mem.offset = int(spec[2])
+        else:
+            mem.offset = int(spec[3])
+
+    def _refresh_setting(self, entry):
+        name = entry.get_name()
+        setting = self._get_setting_data(name)
+        # TODO: It would be nice to bump_wait_dialog here...
+        #       Also, this is really only useful for the cli. :(
```

Comment:
Yeah, unfortunately, .. hmm, what if you make this a generator? I might be able to make the UI iterate that and do the bumping for you as you read the settings out.

Reply:
Let me play with this idea (tomorrow) and if it seems good I'll show you the python magic to make this effectively a coroutine that generates pieces of the list as the UI consumes them, thus avoiding hanging up the UI while you pull these all out.